upgradeable_contract/old_contract/src/test.rs

```diff
@@ -44,3 +45,18 @@ fn test() {
     let client = new_contract::Client::new(&env, &contract_id);
     assert_eq!(1010101, client.new_v2_fn());
 }
+
+#[test]
+fn test_cannot_re_init() {
+    let env = Env::default();
+    env.mock_all_auths();
+
+    // Note that we use register_contract_wasm instead of register_contract
+    // because the old contracts WASM is expected to exist in storage.
+    let contract_id = env.register_contract_wasm(None, old_contract::WASM);
+
+    let client = old_contract::Client::new(&env, &contract_id);
+    let admin = Address::generate(&env);
+    client.init(&admin);
+    assert_eq!(client.try_init(&admin), Err(Ok(Error::AlreadyInitialized)));
```

Comment:
@dmkozh PTAL. The error I got is 

```
❯ cargo test
   Compiling soroban-upgradeable-contract-old-contract v0.0.0 (/Users/matseq/Documents/soroban-playground/2fork-soroban-examples/upgradeable_contract/old_contract)
error[E0308]: mismatched types
  --> src/test.rs:61:41
   |
61 |     assert_eq!(client.try_init(&admin), Err(Ok(Error::AlreadyInitialized)));
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<Result<(), ConversionError>, ...>`, found `Result<_, Result<Error, _>>`
   |
   = note: `Error` and `soroban_sdk::Error` have similar names, but are actually distinct types
```

Reply:
It seems like you just need to remove `use super::*;` (that should be removed in any case) and use `crate::Error` instead.

Reply:
Added `use crate::Error;` and removed `use super::*;`, but still getting a similar error 

```
error[E0308]: mismatched types
  --> src/test.rs:61:41
   |
61 |     assert_eq!(client.try_init(&admin), Err(Ok(Error::AlreadyInitialized)));
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<Result<(), ...>, ...>`, found `Result<_, Result<Error, _>>`
```

Reply:
Sorry, that's indeed pretty confusing. IIUC the client generated from Wasm doesn't use your contract error and returns a generic SDK error instead, while in the other examples the client is generated from the Rust struct directly and has a correct type. So the conversion is necessary to assert on the final error. So something like this works (I realize it's uglier, but I'm not sure we can simplify much more):
```rust
    // `try_init` is expected to return an error. Since client is generated from Wasm,
    // this is a generic SDK error.
    let err: soroban_sdk::Error = client.try_init(&admin).err().unwrap().unwrap();
    // Convert the SDK error to the contract error.
    let contract_err: Error = err.try_into().unwrap();
    // Make sure contract error has the expected value.
    assert_eq!(contract_err, Error::AlreadyInitialized);
```

Reply:
Got it! Took quite a few unwraps 😄 . I decided to leave your comments in the code to serve as an explanatory example of how to do it this way. It should be good to go now. Thanks!

Reply:
Thanks for bearing with me!